### ISSUE TITLE
Reading cast information from RSK file

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#2025]
@dankelley wrote:
> I looked a bit in the files and think I see how to navigate the data to find downcasts and upcasts, but doing so involves reading some tables from the SQL file that have not been read by `read.rsk()` so far.

> My guess is that we look in the `regionCast` table, isolate cases where `type=="DOWN"`, and grab the `regionID` column from that.  Then we match that to the `regionID` column of the `region` table, and extract the `tstamp1` and `tstamp2` values.  Finally, we use those index pairs to look within the `data` slot, which now stores the `tstamp` column, for this purpose.  I'm not sure why I just wrote this paragraph, because the code would be shorter, but I suppose I'd like to get this "on paper" now, for further discussion, before coding.  There may be issues with the fact that the `tstamp` column is kind of weird, being of type `integer64`, which will not work on 32-bit machines ... but frankly I don't see any reason for all users to be denied features because a few users are working on archaic machines.

> Therefore, I added a new arg to `read.rsk()` called `allTables`, which will cause every table to be read and stored in `metadata`.  That way, a user who understands the structure can do whatever they want with the data, e.g. isolating casts, working with location data, etc. At the moment, `allTables` is defaulting to FALSE, but changing that is trivial, of course.

> Below is a demo; click to see ...

<details>

```R
> library(oce)
> f <- "ctd_with_location.rsk"
> d <- read.rsk(f, allTables=TRUE)
Warning messages:
1: In read.rsk(f, allTables = TRUE) :
  FIXME: provide a way to merge metadata$geodata with data
2: In read.rsk(f, allTables = TRUE) :
  to convert RSK timestamp to time: numberAsPOSIXct(as.numeric(tstamp)/1e3,type="unix")
> str(d@data,1)
List of 6
 $ time        : POSIXct[1:17396], format: "2022-08-08 15:09:43.0" ...
 $ conductivity: num [1:17396] -1.63e-04 -3.91e-04 -3.34e-04 -1.78e-04 -2.93e-05 ...
 $ temperature : num [1:17396] 9.38 9.37 9.38 9.4 9.41 ...
 $ pressure    : num [1:17396] 10.1 10.1 10.1 10.1 10.1 ...
 $ turbidity   : num [1:17396] 0.349 0.445 1.85 0.264 1.759 ...
 $ tstamp      :integer64 [1:17396] 1659971383000 1659971383500 1659971384000 1659971384500 1659971385000 1659971385500 1659971386000 1659971386500 ... 
> str(d@metadata,1)
List of 57
 $ units                :List of 4
 $ flags                : list()
 $ filename             : chr "ctd_with_location.rsk"
 $ model                : chr "RBRconcerto3"
 $ pressureType         : chr "absolute"
 $ pressureAtmospheric  : num 10.1
 $ android_metadata     :'data.frame':	1 obs. of  1 variable:
 $ average              :'data.frame':	0 obs. of  5 variables:
 $ breaksdata           :'data.frame':	28 obs. of  2 variables:
 $ burst                :'data.frame':	0 obs. of  5 variables:
 $ caldata              :'data.frame':	0 obs. of  3 variables:
 $ calsource            :'data.frame':	0 obs. of  2 variables:
 $ channels             :'data.frame':	9 obs. of  9 variables:
 $ chartView            :'data.frame':	15 obs. of  3 variables:
 $ continuous           :'data.frame':	0 obs. of  3 variables:
 $ data                 :'data.frame':	17396 obs. of  10 variables:
 $ datasetDeployments   :'data.frame':	0 obs. of  2 variables:
 $ datasets             :'data.frame':	0 obs. of  2 variables:
 $ dbInfo               :'data.frame':	2 obs. of  2 variables:
 $ deployments          :'data.frame':	1 obs. of  10 variables:
 $ directional          :'data.frame':	0 obs. of  7 variables:
 $ downloadprogress     :'data.frame':	1 obs. of  3 variables:
 $ downloads            :'data.frame':	2215 obs. of  4 variables:
 $ downloadsevents      :'data.frame':	63 obs. of  4 variables:
 $ downloadsheader      :'data.frame':	1 obs. of  4 variables:
 $ downsample_caches    :'data.frame':	1 obs. of  2 variables:
 $ epochs               :'data.frame':	1 obs. of  4 variables:
 $ errors               :'data.frame':	0 obs. of  5 variables:
 $ events               :'data.frame':	204 obs. of  5 variables:
 $ geodata              :'data.frame':	15491 obs. of  6 variables:
 $ instrumentChannels   :'data.frame':	9 obs. of  4 variables:
 $ instruments          :'data.frame':	1 obs. of  2 variables:
 $ parameters           :'data.frame':	1 obs. of  2 variables:
 $ power                :'data.frame':	0 obs. of  13 variables:
 $ regimes              :'data.frame':	0 obs. of  5 variables:
 $ region               :'data.frame':	111 obs. of  8 variables:
 $ regionCal            :'data.frame':	0 obs. of  4 variables:
 $ regionCast           :'data.frame':	46 obs. of  3 variables:
 $ regionComment        :'data.frame':	0 obs. of  1 variable:
 $ regionExclude        :'data.frame':	0 obs. of  3 variables:
 $ regionGeoData        :'data.frame':	42 obs. of  3 variables:
 $ regionPlateau        :'data.frame':	0 obs. of  4 variables:
 $ regionProfile        :'data.frame':	23 obs. of  1 variable:
 $ schedules            :'data.frame':	1 obs. of  8 variables:
 $ sqlite_sequence      :'data.frame':	0 obs. of  2 variables:
 $ thumbnailData        :'data.frame':	0 obs. of  5 variables:
 $ tide                 :'data.frame':	0 obs. of  5 variables:
 $ valve                :'data.frame':	0 obs. of  10 variables:
 $ valveSchedule        :'data.frame':	0 obs. of  7 variables:
 $ valveScheduleSegments:'data.frame':	0 obs. of  5 variables:
 $ wave                 :'data.frame':	0 obs. of  5 variables:
 $ wifi                 :'data.frame':	0 obs. of  3 variables:
 $ dataNamesOriginal    : Named chr [1:6] "-" "cond19" "temp14" "pres24" ...
  ..- attr(*, "names")= chr [1:6] "time" "conductivity" "temperature" "pressure" ...
 $ serialNumber         : int 209642
 $ sampleInterval       : num 0.5
 $ rskVersion           : num [1:3] 1 3 0
 $ ruskinVersion        : chr "mobile"
```

_Originally posted by @dankelley in https://github.com/dankelley/oce/issues/2024#issuecomment-1345584931_

**Comment by dankelley** [comment]
I tested the scheme noted in the comment (code after Details; see the last few lines).

The RBR scheme seems quite good, although if I were doing this by hand (or by a scheme that is not being discussed in a public GH repo yet) then it would have been tighter to the region where P was more clearly on an increasing trend.

![rsk_with_location_05](https://user-images.githubusercontent.com/99469/207029443-6e31ce94-1d36-4001-a7ea-596a343d7f08.png)

<details>

```R
library(oce)
source("~/git/oce/R/rsk.R")
f <- "ctd_with_location.rsk"
d <- read.rsk(f, allTables=TRUE)
geodata <- d[["geodata"]]
geodata$time <- numberAsPOSIXct(geodata$tstamp/1e3,type="unix")-8*3600
range(geodata$time)
time <- d[["time"]]
range(time)
lat <- approx(geodata$time, geodata$latitude, time)$y
lon <- approx(geodata$time, geodata$longitude, time)$y
d <- oceSetData(d, "latitude", lat)
d <- oceSetData(d, "longitude", lon)

# Map colour-coded by time
if (!interactive())
    png("rsk_with_location_%02d.png")
par(mar=c(2,2,2,2), mgp=c(2,0.75,0))
cm <- colormap(time)
drawPalette(colormap=cm)
asp <- 1/cos(pi/180*mean(range(d[["latitude"]])))
plot(d[["longitude"]], d[["latitude"]],
    xlab="", ylab="", asp=asp, bg=cm$zcol, pch=21)
data(coastlineWorldFine, package="ocedata")
polygon(coastlineWorldFine[["longitude"]],
    coastlineWorldFine[["latitude"]],
    col=rgb(210/255,180/255,140/255,alpha=0.2))

k <- kmeans(cbind(lon,lat), centers=30)
points(k$centers[,1], k$centers[,2], col=2, pch="+", cex=2)

plot(lon, type="o", cex=0.3) # ~22 points, by eye
plot(lat, type="o", cex=0.3) # ~18 points, by eye
#smoothScatter(lon,lat,bandwidth=0.001*c(asp,1)) # 23 points, by eye
plot(lon, lat, asp=asp, cex=0.3)

regionCast <- d[["regionCast"]]
region <- d[["region"]]
str(region)
rID <- subset(regionCast, type=="DOWN")$regionID
r <- subset(region, regionID %in% rID)
start <- numberAsPOSIXct(r$tstamp1/1e3, type="unix")
end <- numberAsPOSIXct(r$tstamp2/1e3, type="unix")

oce.plot.ts(d[["time"]], d[["pressure"]], xlim=d[["time"]][1]+c(0,800))
abline(v=start, col=3)
abline(v=end, col=2)


```

**Comment by dankelley** [comment]
Here's what I get with a very simple downcast-isolation scheme.  In this particular case, I prefer it to the RBR scheme, because it cuts out "plateau" data.

However, I am inclined to assume that the RBR has been devised to work with a wide range of data, e.g. handling wave noise well, so it may be superior in general.  At any rate, it *is* a scheme that oce users will likely want to investigate, so certainly we ought to have something in oce to handle this, I suppose in `ctdFindProfiles`, which could check whether the object holds these extra tables.

Perhaps in a Z meeting, @richardsc and I can discuss some choices in this.  I am now thinking that my flag for reading all the tables in an RSK file ought to default to TRUE.  This particular object in R is no bigger than the rsk file, so it's not like we will be using too much memory to be practical.

I could do a Z today but the rest of the week is tight.

![rsk_with_location_05](https://user-images.githubusercontent.com/99469/207032459-7ff67506-f663-4cee-a0de-c7a4634de7bd.png)

**Comment by dankelley** [comment]
In the below, the solid vert lines are from the RBR suggestions, and the dotted vert ones are from a simple method of my own.

![rsk_with_location_6](https://user-images.githubusercontent.com/99469/207085033-d7c22464-9bd0-4341-88b8-796fed5f324f.png)

**Comment by dankelley** [comment]
I think this issue has been addressed, so I am closing this, on the assumption that @richardsc will reopen it (or create a new issue with a more detailed title) if required.  I'm sorry this took so long.

The "rsk" branch has been merged into the "develop" branch (commit c3bb3fd3e517cc2cd94fbd288ca1767a2a1ec839) and it is the latter that ought to be used for further tests.  

If large-scale recoding is required to address any issues, I'll revive "rsk" but for now, I ask @Brendan-Higgins and @richardsc (the two most involved in the RSK work over the past few weeks) to focus on "develop".  Also, I ask that they create new issues with narrow titles, if new problems arise or new features are desired.  Issue drift gets confusing on github, owing to the lack of threaded comments.

Thanks for the patience on this.

Dan.
